src/irc_bridge.py

```diff
@@ -0,0 +1,165 @@
+        # pylint:disable=duplicate-code
+        # the code that is duplicated checks that Synapse is ready
+        # the relevant lines are three and refactoring this in a separate method
+        # would not add any value
+        """Collect status event handler.
+
+        Args:
+            event: Collect status event.
+            charm_state: Instance of CharmState.
+        """
+        if not charm_state.synapse_config.enable_irc_bridge:
+            return
+        if charm_state.irc_bridge_datasource is None:
+            self._charm.unit.status = ops.MaintenanceStatus("Waiting for irc bridge db relation")
+            return
+        container = self._charm.unit.get_container(synapse.SYNAPSE_CONTAINER_NAME)
+        if not container.can_connect():
+            self._charm.unit.status = ops.MaintenanceStatus("Waiting for Synapse pebble")
+            return
+        irc_service = container.get_services(IRC_SERVICE_NAME)
+        if irc_service:
+            logger.debug("%s service already exists, skipping", IRC_SERVICE_NAME)
+            return
+        self._enable_irc_bridge(charm_state)
+        event.add_status(ops.ActiveStatus())
+
+    def _enable_irc_bridge(self, charm_state: CharmState) -> None:
+        """Enable irc service.
+
+        The required steps to enable the IRC bridge are:
+         - Create the IRC bridge configuration file.
+         - Create the IRC bridge registration file.
+         - Generate a PEM file for the IRC bridge.
+         - Finally, add IRC bridge pebble layer.
+
+        Args:
+            charm_state: Instance of CharmState.
+
+        """
+        container = self._charm.unit.get_container(synapse.SYNAPSE_CONTAINER_NAME)
+        if not container.can_connect():
+            self._charm.unit.status = ops.MaintenanceStatus("Waiting for Synapse pebble")
+            return
+        self._charm.model.unit.status = ops.MaintenanceStatus("Configuring IRC bridge")
+        db_connect_string = self._get_db_connection(charm_state)
+        synapse.create_irc_bridge_config(
+            container=container, charm_state=charm_state, db_connect_string=db_connect_string
+        )
+        synapse.create_irc_bridge_app_registration(container=container)
+        self._create_pem_file(container=container)
+        pebble.replan_irc_bridge(container)
+        pebble.restart_synapse(charm_state, container)
+        self._charm.model.unit.status = ops.ActiveStatus()
+
+    def _create_pem_file(self, container: ops.model.Container) -> None:
+        """Create a PEM file for the IRC bridge.
+
+        Args:
+            container: The container to create the PEM file in.
+
+        Raises:
+            PEMCreateError: If the PEM file creation fails.
+        """
+        self._charm.model.unit.status = ops.MaintenanceStatus("Creating PEM file for IRC bridge")
+        pem_create_command = [
+            "/bin/bash",
+            "-c",
+            "[[ -f /data/config/irc_passkey.pem ]] || "
```

Comment:
Heads up: I believe the /data/config is created only if Mjolnir is enabled.
https://github.com/canonical/synapse-operator/blob/main/src/synapse/workload.py#L543